identity/v2/service.go

```diff
@@ -0,0 +1,306 @@
+		s.log.Errorf("Couldn't create account ID: %s", err)
+
+		return ErrAccountIDCreation
+	}
+
+	if err := s.ValidateAccount(ctx, accID); err != nil {
+		s.log.Errorf("Invalid identity - %s: %s", accID.ToHexString(), err)
+
+		return ErrInvalidAccount
+	}
+
+	if _, err := url.ParseRequestURI(req.WebhookURL); err != nil {
+		s.log.Errorf("Invalid webhook URL: %s", err)
+
+		return ErrInvalidWebhookURL
+	}
+
+	return nil
+}
+
+func (s *service) ValidateKey(
+	ctx context.Context,
+	accountID *types.AccountID,
+	pubKey []byte,
+	keyPurpose keystoreType.KeyPurpose,
+) error {
+	// TODO(cdamian): Add validation from the NFT branch
+
+	keyID := &keystoreType.KeyID{
+		Hash:       types.NewHash(pubKey),
+		KeyPurpose: keyPurpose,
+	}
+
+	account, err := s.configService.GetAccount(accountID.ToBytes())
+
+	if err != nil {
+		s.log.Errorf("Couldn't retrieve account: %s", err)
+
+		return ErrAccountRetrieval
+	}
+
+	ctx = contextutil.WithAccount(ctx, account)
+
+	key, err := s.keystoreAPI.GetKey(ctx, keyID)
+
+	if err != nil {
+		s.log.Errorf("Couldn't retrieve key: %s", err)
+
+		return ErrKeyRetrieval
+	}
+
+	if key == nil {
+		s.log.Errorf("Key not found")
+
+		return ErrKeyNotFound
+	}
+
+	if !key.RevokedAt.HasValue() {
+		return nil
+	}
+
+	ok, revokedAt := key.RevokedAt.Unwrap()
+
+	if !ok {
+		s.log.Errorf("Invalid key data - revoked block number should be present")
+
+		return ErrInvalidKeyData
+	}
+
+	latestBlock, err := s.centAPI.GetBlockLatest()
+
+	if err != nil {
+		s.log.Errorf("Couldn't retrieve latest block")
+
+		return ErrLatestBlockRetrieval
+	}
+
+	if types.U32(latestBlock.Block.Header.Number) >= revokedAt {
+		s.log.Errorf("Key is revoked")
+
+		return ErrKeyRevoked
+	}
+
+	return nil
+}
+
+func (s *service) ValidateSignature(
+	ctx context.Context,
+	accountID *types.AccountID,
+	pubKey []byte,
+	message []byte,
+	signature []byte,
+) error {
+	if err := s.ValidateKey(ctx, accountID, pubKey, keystoreType.KeyPurposeP2PDocumentSigning); err != nil {
+		s.log.Errorf("Couldn't validate key: %s", err)
+
+		return err
+	}
+
+	if !crypto.VerifyMessage(pubKey, message, signature, crypto.CurveEd25519) {
+		s.log.Error("Couldn't verify message - invalid signature")
+
+		return ErrInvalidSignature
+	}
+
+	return nil
+}
+
+func (s *service) ValidateAccount(_ context.Context, accountID *types.AccountID) error {
+	meta, err := s.centAPI.GetMetadataLatest()
+
+	if err != nil {
+		s.log.Errorf("Couldn't get latest metadata: %s", err)
+
+		return ErrMetadataRetrieval
+	}
+
+	accountStorageKey, err := types.CreateStorageKey(meta, "System", "Account", accountID.ToBytes())
+
+	if err != nil {
+		s.log.Errorf("Couldn't create storage key for account: %s", err)
+
+		return ErrAccountStorageKeyCreation
+	}
+
+	var accountInfo types.AccountInfo
+
+	ok, err := s.centAPI.GetStorageLatest(accountStorageKey, &accountInfo)
+
+	if err != nil {
+		s.log.Errorf("Couldn't retrieve account from storage: %s", err)
+
+		return ErrAccountStorageRetrieval
+	}
+
+	if !ok {
```

Comment:
Sometimes a proxy will not have any balance and therefore there will not be a record of it in the `System.Account` storage. I think here would be safer to ensure that the AccountID is a registered proxy.

Reply:
> Sometimes a proxy will not have any balance and therefore there will not be a record of it in the System.Account storage.
What about anonymous proxies? Gonna try that locally and see what I get.

> I think here would be safer to ensure that the AccountID is a registered proxy.
We do not have the required information at the level where this is usually called i.e. field validators for documents. In order to check if this is a proxy I would need 2 addresses, delegator and delegate.

Reply:
Most of the identities are going to be anonymous proxies and they may not have an entry on the `System.Account` storage. 
You can check it by calling `proxy.proxies(AccountID)`, only one account needed.

Reply:
Added an extra check in the identities service to address the suggested approach.